pkg/apply/apply.go

```diff
@@ -3,19 +3,84 @@ package apply
 import (
 	"context"
 	"fmt"
+	corev1 "k8s.io/api/core/v1"
 	"log"
+	"reflect"
 
 	"github.com/pkg/errors"
 
+	"gopkg.in/yaml.v2"
 	"k8s.io/apimachinery/pkg/api/equality"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	uns "k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/types"
 	k8sclient "sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-// Find existing object or create if if it doesn't exists
-func findOrCreateObject(ctx context.Context, client k8sclient.Client, obj *uns.Unstructured) (*uns.Unstructured, string, error) {
+type ConfSectionSelector int
+
+const (
+	ADDRESSPOOL ConfSectionSelector = iota
+	BGPPEER
+)
+
+func UpdateConfigMapObjs(ctx context.Context, client k8sclient.Client, objs []*uns.Unstructured, confSel ConfSectionSelector, namespace string) error {
+	cur, upd := ConfigMapData{}, ConfigMapData{}
+	var configMap corev1.ConfigMap
+
+	if err := client.Get(ctx, k8sclient.ObjectKey{Name: MetalLBConfigMap, Namespace: namespace}, &configMap); err != nil {
```

Comment:
What happens if the configmap is not there? IIRC we start with no configmap (plus, there are some cases where this function deletes it). Should this go on if the error is not found?

And should let the error float otherwise?

Reply:
that is why I have update and create path decoupled so you will always comes here with configmap avaliable

Reply:
Not sure I follow. Let's say we start with empty.

You enter with no address pools and end up here 
https://github.com/msherif1234/metallb-operator/blob/f4ef087028415444b02a101a4a7423bd39b15cb2/controllers/addresspool_controller.go#L62

That goes into `UpdateConfigMapObjs` and it exits because the cm is not there.
Then, we create the bgp peer object, which again ends up into `UpdateConfigMapObjs` which does not find the cm and exists.

What am I missing? And in any case, having this check does not justify the extra code complexity.

Reply:
hmm create path doesn't use the updateonfigmapObjs ?

Reply:
Again, I don't get it. The create path is when the get fails with not found, here https://github.com/msherif1234/metallb-operator/blob/f4ef087028415444b02a101a4a7423bd39b15cb2/controllers/addresspool_controller.go#L61

And that uses `UpdateConfigMapObjs`

Reply:
that isn't the 1st time create path the create path is starting from line 69

Reply:
Checked offline, now I understand, sorry. 
Still, I'd use the same path for both the actions.

Reply:
not sure what is left here as valid comments and what is not I have been looking at merging create and delete code path and they are just completly different code apis etc I don't see any value from using same path for both action, they were already different originally , also pulling logic out of updateconfig and moving it to controller also doesn't sound like a good idea the whole idea of using apply pkg to abstract and do the heavy lifting there and keep the controller light , I am with any optimization if it makes sense but I just don't see it

Reply:
This comment about marshalling on one end and unmarshalling on the other https://github.com/metallb/metallb-operator/pull/71#discussion_r703659793 

I don't think it makes much sense